static/flow/views/program-editor-panel.js

```diff
@@ -91,40 +104,108 @@ var ProgramEditorPanel = function(options) {
+    //
+    var storeProgramState = function(overwrite) {
+        if (overwrite) {
+            _this.programSpecStored[_this.programSpecStored.length - 1] = diagramToSpec(_this.m_diagram);
+        } else {
+            _this.programSpecStored.push(diagramToSpec(_this.m_diagram));
         }
+        if (_this.programSpecStored.length > _this.maxSpecHistory) {
+            _this.programSpecStored.shift();
+        }
+    };
 
-        //
-        // Display connections
-        //
-        for (var i = 0; i < this.m_diagram.blocks.length; i++) {
-            var block = this.m_diagram.blocks[i];
-            for (var j = 0; j < block.pins.length; j++) {
-                var pin = block.pins[j];
-                if (pin.sourcePin) {
-                    // console.log("[DEBUG] displayConnection", pin);
-                    this.displayConnection(pin, this.m_scale);
-                }
-            }
+    //
+    // use ctrl+ z to undo action in editor
+    //
+    var undoChange = function() {
+
+        // see if we have a valid spec in the storage array that we can undo to
+		// nth block is current state, n-1 entry is first valid undo state
+        if (_this.programSpecStored.length < 2) {
+            return;
+        }
+
+        // clear existing program
+        _this.undisplayAllBlocks();
+        _this.nameHash = {};
+		
+        // the final spec in the array is the current state, get the (n-1)th entry
+        _this.m_diagram = specToDiagram(_this.programSpecStored[_this.programSpecStored.length - 2]);
+
+        // remove last stored state from array
+        _this.programSpecStored.pop();
+
+        _this.displayAllBlocks();
+
+        _this.displayAllConnections();
+
+		// resave program, but do not resave the state
+        _this.autoSaveProgram(false, false);
+    };
+
+    // bind the undo key when we init, be sure only to call once
+    $(document).bind('keydown', 'ctrl+z', undoChange);
+
+    //
+    // use ctrl+ z to undo action in editor
+    //
+    this.autoSaveProgram = function(moveBlock = false, saveState = true) {
+        updateSaveStatus("Saving...");
+        // lump block moves together
+        // if the last change was a move and this is a move
+        // then only save the most recent move
+        if (saveState) {
+            var overwrite = moveBlock && _this.lastChangeMoveBlock;
+            storeProgramState(overwrite);
+            _this.lastChangeMoveBlock = moveBlock;
+        }
+        if (_this.saving) {
+            _this.saveAgain = true;
+            return;
         }
+        _this.saving = true;
+        saveProgram(saveComplete);
+    };
+
+    //
+    // called when save is complete
+    //
+    var saveComplete = function(success) {
+        _this.saving = false;
+        updateSaveStatus("All changes saved");
+        if (_this.saveAgain) {
+            _this.saveAgain = false;
+            _this.autoSaveProgram(false, false);
```

Comment:
I'm trying to understand - is the idea with `saveAgain` that saving doesn't work if you trigger it during a save, so you need to re-trigger the save here without re-storing the current state? If that's the case, it might be worth clarifying this in some comments.

Reply:
Interesting that you mention it - I thought a redo might be in order too.  I'll add a story.  And, yeah, you nailed the `saveAgain` - since save is asynchronous we might be in the middle of a save when a request one ore more additional saves come in (e.g., user rapidly clicks to add a bunch of blocks).  We set the `saveAgain` flag so that when the initial save is complete, we can fire another one.